### DISCUSSION TITLE
sql injection

### DISCUSSION THREAD

**Discussion by hamza-ashfaq751** [prisma/prisma#19533] [Q&A]
prisma methods create update upsert are safed from sql injection or not  ?

**Comment by ludralph** [comment]
Hi @hamza-ashfaq751 👋 

Prisma's ORM methods, such as create, update, and upsert, are designed to generate parameterized SQL queries. Parameterized queries are considered safe from SQL injection attacks because they separate the SQL code from the data values being used in the query. The values are passed as parameters, which are properly escaped and sanitized by the underlying database driver.

**Reply by ludralph** [reply]
Hi @hamza-ashfaq751 

Do you have any more questions around this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!

**Reply by nurul3101** [reply]
We are closing this Discussion due to inactivity. Please open a new Discussion if you have the a similar problem 🙏

**Reply by Urbaninja25** [reply]
Hello @ludralph ,its nice to see u helping us here ! 🫡

do we have  more Parameterized queries(data sanitation already done out of the box).wtb filtering (findMany or findUnique) .

i cant find in prisma docs high level  queries  which dont need explicit data sanitation  to avoid malicious  sql injections


thank u again !

**Reply by ludralph** [reply]
Hi @Urbaninja25 👋 

Prisma is designed to provide a type-safe API for database access, which inherently helps prevent SQL injection by using parameterized queries. This means that when you use Prisma's methods like `findMany` or `findUnique`, the parameters you pass are automatically sanitized to prevent SQL injection attacks.
For example, when you use `findUnique`, you would provide parameters in a way that Prisma prepares the query safely:
```js
const result = await prisma.user.findUnique({
  where: {
    email: 'alice@prisma.io',
  },
})
```
Similarly, for `findMany`, you would pass the filter conditions as parameters, and Prisma would handle them safely:
```js
const emails = ['alice@example.com', 'bob@example.com'];
user.findMany({ where: { emails: { in: emails } } });
```
To see what the query looks like under the hood, you can enable logging. A sample query log is 
```
prisma:query SELECT "public"."User"."user_id", "public"."User"."full_name", "public"."User"."email_address", "public"."User"."created_at", "public"."User"."updated_at" FROM "public"."User" WHERE "public"."User"."full_name" = $1 OFFSET $2
```
These examples show that Prisma handles the query parameterization for you, which is a key defense against SQL injection. However, it's always good practice to validate and sanitize user input at the application level before it reaches your database queries as an additional layer of security.
